src/nodes.cpp

```diff
@@ -0,0 +1,41 @@
+// NODES
+#include <string>
+
+//#include <eigen3/Eigen>  // for working with vectors and matrices later
+
+class BaseNode {
+
+	private:
+		std::string derivative = "";  // or expression
+
+
+		auto Evaluate() {
+			// evaluate children
+		};
+
+		auto Derive(this) {
+			// find derivative of this node (using children derivatives)
+		};
+
+	public:
+		std::string getDerivative(this) {
+			return derivative;
+		}
+		void setDerivative(this, d) {
+			this.derivative = d;
+		}
+		
+}; // END BaseNode class
+
+class Operation(BaseNode) {
+
+	private:
+		std::string[2] branches;  // branches to 2 children (always 2 children for simplicity)
```

Comment:
Do we want the branches to be strings? I think we should have them as references to instances of BaseNode. This is because I assumed the string parser would convert an expression represented as a string into an expression tree whos root is a BaseNode, meaning that once we have the expression tree we never have to touch the string representation again for doing any calculations.

Reply:
Yes I am unsure. It depends how we want to implement it. I have added an id property to the classes